src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/X509AsymmetricSecurityKey.cs

```diff
@@ -0,0 +1,441 @@
+﻿using System;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography.Xml;
+
+namespace CoreWCF.IdentityModel.Tokens
+{
+    public class X509AsymmetricSecurityKey : AsymmetricSecurityKey
+    {
+        private X509Certificate2 _certificate;
+        private AsymmetricAlgorithm _privateKey;
+        private bool _privateKeyAvailabilityDetermined;
+        private AsymmetricAlgorithm _publicKey;
+        private bool _publicKeyAvailabilityDetermined;
+
+        public X509AsymmetricSecurityKey(X509Certificate2 certificate)
+        {
+            _certificate = certificate ?? throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(certificate));
+        }
+
+        public override int KeySize
+        {
+            get { return PublicKey.KeySize; }
+        }
+
+        private AsymmetricAlgorithm PrivateKey
+        {
+            get
+            {
+                if (!_privateKeyAvailabilityDetermined)
+                {
+                    lock (ThisLock)
+                    {
+                        _privateKey = _certificate.GetRSAPrivateKey();
+                        if (_privateKey != null)
+                        {
+                            RSACryptoServiceProvider rsaCsp = _privateKey as RSACryptoServiceProvider;
+                            // ProviderType == 1 is PROV_RSA_FULL provider type that only supports SHA1. Change it to PROV_RSA_AES=24 that supports SHA2 also.
+                            if (rsaCsp != null && rsaCsp.CspKeyContainerInfo.ProviderType == 1)
+                            {
+                                CspParameters csp = new CspParameters();
+                                csp.ProviderType = 24;
+                                csp.KeyContainerName = rsaCsp.CspKeyContainerInfo.KeyContainerName;
+                                csp.KeyNumber = (int)rsaCsp.CspKeyContainerInfo.KeyNumber;
+                                if (rsaCsp.CspKeyContainerInfo.MachineKeyStore)
+                                {
+                                    csp.Flags = CspProviderFlags.UseMachineKeyStore;
+                                }
+
+                                csp.Flags |= CspProviderFlags.UseExistingKey;
+                                _privateKey = new RSACryptoServiceProvider(csp);
+                            }
+                        }
+                        else
+                        {
+                            _privateKey = _certificate.GetECDsaPrivateKey();
+                            // We don't support DSA certificates as we need DSACertificateExtensions which is in netstandard2.1 and netcore2.0.
```

Comment:
Can you create an issue to enable support for DSA Certificates when we move forward to a later .NET version and mention this method

Reply:
https://github.com/CoreWCF/CoreWCF/issues/255